Log OpenTelemetry Trace ID

**Is your feature request related to a problem? Please describe.**
In order to correlate log items of harbor clients and OpenTelemerty traces to harbor logs, the OpenTelemetry Trace ID should be logged (if it's set).

**Describe the solution you'd like**
If the Trace ID is sent to the Harbor in HTTP header or the tracing is enabled (and the Trace ID is generated), the Trace ID should be added to the log lines as a new field, for example:
```
2022-12-27T14:31:38Z [ERROR] [/server/v2.0/handler/user.go:221][requestID="9f2cae97-4d6f-4977-9209-43ba9e073019" traceID="0af7651916cd43dd8448eb211c80319c"]: Failed to get user info for ID 0, error: user 0 not found
```
Where the `taceID` is the given Trace ID.

**Describe the main design/architecture of your solution**
The log middleware should be extended by adding the Trace ID to the fields of context logger. All operations, which use the context logger, `log.G(ctx)`, will log the Trace ID (only if it's given) as a log field

There are 3 use cases:
a) Tracing is not enabled, OpenTelemetry HTTP header `traceparent` is missing: Nothing happen, the `traceID` field is disappeared in the log lines
b) Tracing is enabled: Trace ID is created by trace middleware, or OpenTelemetry HTTP header `traceparent` is used (by otel trace lib). The Span added to the context contains the Trace ID, which will be extracted for the log middleware. The `traceID` field is appeared in the log lines
c) Tracing is not enabled: If OpenTelemetry HTTP header `traceparent` is set by Harbor client, Trace ID is extracted from the header. The `traceID` field is appeared in the log lines

See more details about using OpenTelemetry HTTP header `traceparent` without any OpenTelemetry collector in my article: <https://medium.com/faun/multi-hop-tracing-with-opentelemetry-in-golang-792df5feb37c>.

**Describe the development plan you've considered**
See pull request  #18181 .

**Additional context**
Example `curl` command to use OpenTelemetry HTTP header `traceparent`:
```
curl -u test:Test1234 -H 'traceparent: 00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01' harbor.kind-01.company.com:88/api/v2.0/users/0
```
The above example code prints out below log line:
```
2022-12-27T14:31:38Z [ERROR] [/server/v2.0/handler/user.go:221][requestID="9f2cae97-4d6f-4977-9209-43ba9e073019" traceID="0af7651916cd43dd8448eb211c80319c"]: Failed to get user info for ID 0, error: user 0 not found
```

--- a/src/lib/trace/helper.go
+++ b/src/lib/trace/helper.go
@@ -22,6 +22,7 @@ import (
 	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
 	"go.opentelemetry.io/otel"
 	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/propagation"
 	oteltrace "go.opentelemetry.io/otel/trace"
 )
 
@@ -85,4 +86,34 @@ func NewHandler(h http.Handler, operation string) http.Handler {
 // StartTrace returns a new span with the given name.
 func StartTrace(ctx context.Context, tracerName string, spanName string, opts ...oteltrace.SpanStartOption) (context.Context, oteltrace.Span) {
 	return otel.Tracer(tracerName).Start(ctx, spanName, opts...)
+}
+
+/*
+ExractTraceID returns OpenTelemetry Trace ID
+
+	If tracing is enabled, the trace middleware already put the Span into request context,
+		so the func returns the Trace ID form request context
+	If tracing is NOT enabled, the func returns the Trace ID from the request header (if exists)
+	If Trace ID is not found, the func returns empty string
+*/
+func ExractTraceID(r *http.Request) string {
+	var traceID string
+	if Enabled() {
+		sc := oteltrace.SpanContextFromContext(r.Context())
+		if sc.HasTraceID() {
+			traceID = sc.TraceID().String()
+		}
+	}
+	if traceID == "" {
+		if tpHeader := r.Header.Get("traceparent"); tpHeader != "" {
+			var prop propagation.TraceContext
+			ctx := prop.Extract(context.Background(), propagation.HeaderCarrier(r.Header))
+			sc := oteltrace.SpanContextFromContext(ctx)
+			if sc.HasTraceID() {
+				traceID = sc.TraceID().String()
+			}
+		}
+	}
+
+	return traceID
 }

--- a/src/lib/trace/helper_test.go
+++ b/src/lib/trace/helper_test.go
@@ -15,9 +15,15 @@
 package trace
 
 import (
+	"context"
 	"net/http"
+	"net/http/httptest"
 	"net/url"
 	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"go.opentelemetry.io/otel/propagation"
 )
 
 func TestHarborSpanNameFormatter(t *testing.T) {
@@ -85,6 +91,207 @@ func TestHarborSpanNameFormatter(t *testing.T) {
 			if got != tt.want {
 				t.Errorf(`(%v) = %v; want "%v"`, tt.in, got, tt.want)
 			}
+		})
+	}
+}
+
+func TestExractTraceID(t *testing.T) {
+	type args struct {
+		headers        map[string]string
+		ctxTraceparent string
+		traceEnabled   bool
+	}
+	tests := []struct {
+		name string
+		args args
+		want string
+	}{
+		{
+			name: "Dummy",
+			args: args{
+				headers: map[string]string{},
+			},
+			want: "",
+		},
+		{
+			name: "Traceparent Header, trace enabled",
+			args: args{
+				headers: map[string]string{
+					"traceparent": "00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01",
+				},
+				traceEnabled: true,
+			},
+			want: "0af7651916cd43dd8448eb211c80319c",
+		},
+		{
+			name: "Traceparent Header, trace not enabled",
+			args: args{
+				headers: map[string]string{
+					"traceparent": "00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01",
+				},
+				traceEnabled: false,
+			},
+			want: "0af7651916cd43dd8448eb211c80319c",
+		},
+		{
+			name: "Traceparent Header, invalid",
+			args: args{
+				headers: map[string]string{
+					"traceparent": "INVALID",
+				},
+			},
+			want: "",
+		},
+		{
+			name: "Traceparent Context, trace enabled",
+			args: args{
+				headers:        map[string]string{},
+				ctxTraceparent: "00-80e1afed08e019fc1110464cfa66635c-7a085853722dc6d2-01",
+				traceEnabled:   true,
+			},
+			want: "80e1afed08e019fc1110464cfa66635c",
+		},
+		{
+			name: "Traceparent Context, trace not enabled",
+			args: args{
+				headers:        map[string]string{},
+				ctxTraceparent: "00-80e1afed08e019fc1110464cfa66635c-7a085853722dc6d2-01",
+				traceEnabled:   false,
+			},
+			want: "",
+		},
+		{
+			name: "Traceparent Context, invalid, trace enabled",
+			args: args{
+				headers:        map[string]string{},
+				ctxTraceparent: "INVALID",
+				traceEnabled:   true,
+			},
+			want: "",
+		},
+		{
+			name: "Traceparent Context, invalid, trace not enabled",
+			args: args{
+				headers:        map[string]string{},
+				ctxTraceparent: "INVALID",
+				traceEnabled:   false,
+			},
+			want: "",
+		},
+		{
+			name: "Traceparent Context+Header, trace enabled",
+			args: args{
+				headers: map[string]string{
+					"traceparent": "00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01",
+				},
+				ctxTraceparent: "00-80e1afed08e019fc1110464cfa66635c-7a085853722dc6d2-01",
+				traceEnabled:   true,
+			},
+			want: "80e1afed08e019fc1110464cfa66635c",
+		},
+		{
+			name: "Traceparent Context+Header, trace not enabled",
+			args: args{
+				headers: map[string]string{
+					"traceparent": "00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01",
+				},
+				ctxTraceparent: "00-80e1afed08e019fc1110464cfa66635c-7a085853722dc6d2-01",
+				traceEnabled:   false,
+			},
+			want: "0af7651916cd43dd8448eb211c80319c",
+		},
+		{
+			name: "Traceparent Context+Header, invalid #1, trace enabled",
+			args: args{
+				headers: map[string]string{
+					"traceparent": "INVALID",
+				},
+				ctxTraceparent: "00-80e1afed08e019fc1110464cfa66635c-7a085853722dc6d2-01",
+				traceEnabled:   true,
+			},
+			want: "80e1afed08e019fc1110464cfa66635c",
+		},
+		{
+			name: "Traceparent Context+Header, invalid #1, trace not enabled",
+			args: args{
+				headers: map[string]string{
+					"traceparent": "INVALID",
+				},
+				ctxTraceparent: "00-80e1afed08e019fc1110464cfa66635c-7a085853722dc6d2-01",
+				traceEnabled:   false,
+			},
+			want: "",
+		},
+		{
+			name: "Traceparent Context+Header, invalid #2, trace enabled",
+			args: args{
+				headers: map[string]string{
+					"traceparent": "00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01",
+				},
+				ctxTraceparent: "INVALID",
+				traceEnabled:   true,
+			},
+			want: "0af7651916cd43dd8448eb211c80319c",
+		},
+		{
+			name: "Traceparent Context+Header, invalid #2, trace not enabled",
+			args: args{
+				headers: map[string]string{
+					"traceparent": "00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01",
+				},
+				ctxTraceparent: "INVALID",
+				traceEnabled:   false,
+			},
+			want: "0af7651916cd43dd8448eb211c80319c",
+		},
+		{
+			name: "Traceparent Context+Header, invalid #3, trace enabled",
+			args: args{
+				headers: map[string]string{
+					"traceparent": "INVALID",
+				},
+				ctxTraceparent: "INVALID",
+				traceEnabled:   true,
+			},
+			want: "",
+		},
+		{
+			name: "Traceparent Context+Header, invalid #3, trace not enabled",
+			args: args{
+				headers: map[string]string{
+					"traceparent": "INVALID",
+				},
+				ctxTraceparent: "INVALID",
+				traceEnabled:   false,
+			},
+			want: "",
+		},
+	}
+
+	origEnabled := C.Enabled
+	defer func() {
+		C.Enabled = origEnabled
+	}()
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			C.Enabled = tt.args.traceEnabled
+
+			ctx := context.Background()
+			if tt.args.ctxTraceparent != "" {
+				var prop propagation.TraceContext
+				ctx = prop.Extract(ctx, propagation.MapCarrier{"traceparent": tt.args.ctxTraceparent})
+			}
+
+			req := httptest.NewRequest("GET", "/v1/library/photon/manifests/2.0", nil).WithContext(ctx)
+			for h, v := range tt.args.headers {
+				req.Header.Set(h, v)
+			}
+
+			traceID := ExractTraceID(req)
+
+			assert.Equal(t, tt.want, traceID, tt.name)
 		})
 	}
 }

--- a/src/server/middleware/log/log.go
+++ b/src/server/middleware/log/log.go
@@ -18,6 +18,7 @@ import (
 	"net/http"
 
 	"github.com/goharbor/harbor/src/lib/log"
+	tracelib "github.com/goharbor/harbor/src/lib/trace"
 	"github.com/goharbor/harbor/src/server/middleware"
 )
 
@@ -30,9 +31,15 @@ func Middleware() func(http.Handler) http.Handler {
 			logger.Debugf("attach request id %s to the logger for the request %s %s", rid, r.Method, r.URL.Path)
 
 			ctx := log.WithLogger(r.Context(), logger.WithFields(log.Fields{"requestID": rid}))
-			next.ServeHTTP(w, r.WithContext(ctx))
-		} else {
-			next.ServeHTTP(w, r)
+			r = r.WithContext(ctx)
 		}
+
+		traceID := tracelib.ExractTraceID(r)
+		if traceID != "" {
+			ctx := log.WithLogger(r.Context(), log.G(r.Context()).WithFields(log.Fields{"traceID": traceID}))
+			r = r.WithContext(ctx)
+		}
+
+		next.ServeHTTP(w, r)
 	})
 }

--- a/src/server/middleware/log/log_test.go
+++ b/src/server/middleware/log/log_test.go
@@ -15,52 +15,192 @@
 package log
 
 import (
+	"bytes"
+	"context"
+	"fmt"
 	"net/http"
 	"net/http/httptest"
+	"regexp"
 	"testing"
 
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/suite"
+	"go.opentelemetry.io/otel/propagation"
 
 	"github.com/goharbor/harbor/src/lib/log"
+	tracelib "github.com/goharbor/harbor/src/lib/trace"
 )
 
 type MiddlewareTestSuite struct {
 	suite.Suite
 }
 
-func (suite *MiddlewareTestSuite) TestMiddleware() {
+func (s *MiddlewareTestSuite) TestTableMiddleware() {
 	next := func(fields log.Fields) http.Handler {
 		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-			log.G(r.Context()).WithFields(fields).Info("this is message")
+			log.G(r.Context()).WithFields(fields).Info("this is message") // variable loc below refers to this line
 
 			w.WriteHeader(http.StatusOK)
 		})
 	}
+	loc := "/server/middleware/log/log_test.go:41"
+	locPrefix := regexp.MustCompile(fmt.Sprintf(`\[([^\s]*)%s\]`, loc))
 
-	{
-		req := httptest.NewRequest("GET", "/v1/library/photon/manifests/2.0", nil)
-		rr := httptest.NewRecorder()
-
-		Middleware()(next(nil)).ServeHTTP(rr, req)
+	type args struct {
+		headers        map[string]string
+		fields         map[string]interface{}
+		ctxTraceparent string
+	}
+	tests := []struct {
+		name string
+		args args
+		want string
+	}{
+		{
+			name: "Dummy",
+			args: args{
+				headers: map[string]string{},
+			},
+			want: fmt.Sprintf("TIMESTAMP [INFO] [%s]%s", loc,
+				"[TestCase=\"Dummy\"]: this is message\n"),
+		},
+		{
+			name: "X-Request-ID",
+			args: args{
+				headers: map[string]string{
+					"X-Request-ID": "fd6139e6-9092-4181-9220-42d3d48bf658",
+				},
+			},
+			want: fmt.Sprintf("TIMESTAMP [INFO] [%s]%s", loc,
+				"[TestCase=\"X-Request-ID\" requestID=\"fd6139e6-9092-4181-9220-42d3d48bf658\"]: this is message\n"),
+		},
+		{
+			name: "X-Request-ID, field",
+			args: args{
+				headers: map[string]string{
+					"X-Request-ID": "fd6139e6-9092-4181-9220-42d3d48bf658",
+				},
+				fields: log.Fields{"method": "GET"},
+			},
+			want: fmt.Sprintf("TIMESTAMP [INFO] [%s]%s", loc,
+				"[TestCase=\"X-Request-ID, field\" method=\"GET\" requestID=\"fd6139e6-9092-4181-9220-42d3d48bf658\"]: this is message\n"),
+		},
+		{
+			name: "Traceparent Header",
+			args: args{
+				headers: map[string]string{
+					"traceparent": "00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01",
+				},
+			},
+			want: fmt.Sprintf("TIMESTAMP [INFO] [%s]%s", loc,
+				"[TestCase=\"Traceparent Header\" traceID=\"0af7651916cd43dd8448eb211c80319c\"]: this is message\n"),
+		},
+		{
+			name: "Traceparent Context",
+			args: args{
+				headers:        map[string]string{},
+				ctxTraceparent: "00-80e1afed08e019fc1110464cfa66635c-7a085853722dc6d2-01",
+			},
+			want: fmt.Sprintf("TIMESTAMP [INFO] [%s]%s", loc,
+				"[TestCase=\"Traceparent Context\" traceID=\"80e1afed08e019fc1110464cfa66635c\"]: this is message\n"),
+		},
+		{
+			name: "Traceparent Context+Header",
+			args: args{
+				headers: map[string]string{
+					"traceparent": "00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01",
+				},
+				ctxTraceparent: "00-80e1afed08e019fc1110464cfa66635c-7a085853722dc6d2-01",
+			},
+			want: fmt.Sprintf("TIMESTAMP [INFO] [%s]%s", loc,
+				"[TestCase=\"Traceparent Context+Header\" traceID=\"80e1afed08e019fc1110464cfa66635c\"]: this is message\n"),
+		},
+		{
+			name: "Traceparent Context+Header, X-Request-ID",
+			args: args{
+				headers: map[string]string{
+					"X-Request-ID": "fd6139e6-9092-4181-9220-42d3d48bf658",
+					"traceparent":  "00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01",
+				},
+				ctxTraceparent: "00-80e1afed08e019fc1110464cfa66635c-7a085853722dc6d2-01",
+			},
+			want: fmt.Sprintf("TIMESTAMP [INFO] [%s]%s", loc,
+				"[TestCase=\"Traceparent Context+Header, X-Request-ID\" requestID=\"fd6139e6-9092-4181-9220-42d3d48bf658\" traceID=\"80e1afed08e019fc1110464cfa66635c\"]: this is message\n"),
+		},
+		{
+			name: "Traceparent Context+Header, X-Request-ID, field",
+			args: args{
+				headers: map[string]string{
+					"X-Request-ID": "fd6139e6-9092-4181-9220-42d3d48bf658",
+					"traceparent":  "00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01",
+				},
+				ctxTraceparent: "00-80e1afed08e019fc1110464cfa66635c-7a085853722dc6d2-01",
+				fields:         log.Fields{"method": "GET"},
+			},
+			want: fmt.Sprintf("TIMESTAMP [INFO] [%s]%s", loc,
+				"[TestCase=\"Traceparent Context+Header, X-Request-ID, field\" method=\"GET\" requestID=\"fd6139e6-9092-4181-9220-42d3d48bf658\" traceID=\"80e1afed08e019fc1110464cfa66635c\"]: this is message\n"),
+		},
 	}
 
-	{
-		req := httptest.NewRequest("GET", "/v1/library/photon/manifests/2.0", nil)
-		req.Header.Set("X-Request-ID", "fd6139e6-9092-4181-9220-42d3d48bf658")
-		rr := httptest.NewRecorder()
+	origEnabled := tracelib.C.Enabled
+	defer func() {
+		tracelib.C.Enabled = origEnabled
+	}()
+	tracelib.C.Enabled = true
 
-		Middleware()(next(nil)).ServeHTTP(rr, req)
-	}
+	for _, tt := range tests {
+		tt := tt
+		s.T().Run(tt.name, func(t *testing.T) {
+			b := make([]byte, 0, 200)
+			buf := bytes.NewBuffer(b)
+			formatter := log.NewTextFormatter()
+			formatter.SetTimeFormat("TIMESTAMP")
+			logger := log.New(buf, formatter, log.InfoLevel, 3).WithField("TestCase", tt.name)
+			ctx := log.WithLogger(context.Background(), logger)
+			if tt.args.ctxTraceparent != "" {
+				var prop propagation.TraceContext
+				ctx = prop.Extract(ctx, propagation.MapCarrier{"traceparent": tt.args.ctxTraceparent})
+			}
 
-	{
-		req := httptest.NewRequest("GET", "/v1/library/photon/manifests/2.0", nil)
-		req.Header.Set("X-Request-ID", "fd6139e6-9092-4181-9220-42d3d48bf658")
-		rr := httptest.NewRecorder()
+			req := httptest.NewRequest("GET", "/v1/library/photon/manifests/2.0", nil).WithContext(ctx)
+			for h, v := range tt.args.headers {
+				req.Header.Set(h, v)
+			}
+			rr := httptest.NewRecorder()
 
-		Middleware()(next(log.Fields{"method": req.Method})).ServeHTTP(rr, req)
+			Middleware()(next(tt.args.fields)).ServeHTTP(rr, req)
+
+			line := string(removeSubmatch(locPrefix, buf.Bytes()))
+			s.Equal(tt.want, line, tt.name)
+		})
 	}
 }
 
 func TestMiddlewareTestSuite(t *testing.T) {
 	suite.Run(t, &MiddlewareTestSuite{})
+}
+
+func removeSubmatch(matchRe *regexp.Regexp, line []byte) []byte {
+	matches := matchRe.FindSubmatchIndex(line)
+	if len(matches) < 4 {
+		return line
+	}
+
+	return append(line[0:matches[2]], line[matches[3]:]...)
+}
+
+func TestRemoveSubmatch(t *testing.T) {
+	loc := "/server/middleware/log/log_test.go:41"
+	locPrefix := regexp.MustCompile(fmt.Sprintf(`\[([^\s]*)%s\]`, loc))
+
+	line := `TIMESTAMP [INFO] [/github.com/goharbor/harbor/src/server/middleware/log/log_test.go:41][method="GET" requestID="fd6139e6-9092-4181-9220-42d3d48bf658" traceID="80e1afed08e019fc1110464cfa66635c"]: this is message`
+	assert.Equal(t, `TIMESTAMP [INFO] [/server/middleware/log/log_test.go:41][method="GET" requestID="fd6139e6-9092-4181-9220-42d3d48bf658" traceID="80e1afed08e019fc1110464cfa66635c"]: this is message`,
+		string(removeSubmatch(locPrefix, []byte(line))),
+	)
+
+	line = `TIMESTAMP [INFO] [/server/middleware/log/log_test.go:41][method="GET" requestID="fd6139e6-9092-4181-9220-42d3d48bf658" traceID="80e1afed08e019fc1110464cfa66635c"]: this is message`
+	assert.Equal(t, `TIMESTAMP [INFO] [/server/middleware/log/log_test.go:41][method="GET" requestID="fd6139e6-9092-4181-9220-42d3d48bf658" traceID="80e1afed08e019fc1110464cfa66635c"]: this is message`,
+		string(removeSubmatch(locPrefix, []byte(line))),
+	)
+
 }

